### PR TITLE
Fix IsSafeToChangeAssociativity for dynamic variables

### LINKED ISSUE #47365: IDE0047 false positive in dynamic expression
**Version Used**:

Microsoft Visual Studio Enterprise 2019
Version 16.7.2, 16.8.0 Preview 2.1

**Steps to Reproduce**:

Enter the following code in the code editor:

```c#
using System;

namespace ConsoleApp2
{
    internal class Program
    {
        private static void Main()
        {
            dynamic i = 1;
            dynamic s = "s";
            Console.WriteLine(s + (1 + i)); // <= IDE0047 reported
        }
    }
}
```

**Expected Behavior**:

No suggestion.

**Actual Behavior**:

IDE0047 is reported suggesting removal of the parentheses. This is a false positive, because doing so changes the meaning of the code (`s11` is printed instead of `s2`).

afbeelding

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #47365

### CODE REVIEW

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs:1]: It looks like the original file didn't have BOM character and Visual Studio added it automatically for me.

@sharwell FYI.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 2]
```diff
@@ -40,6 +40,12 @@ internal static class CommonParenthesizedExpressionSyntaxExtensions
                 }
             }
 
+            // See https://github.com/dotnet/roslyn/issues/47365
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ParenthesizedExpressionSyntaxExtensions.cs:43]: i would prefer an actual comment here.  having a link is fine as well.

Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ParenthesizedExpressionSyntaxExtensions.cs:43]: @CyrusNajmabadi Like this?

```suggestion
            // It's not safe to change associativity for dynamic variables as the actual type isn't known. See https://github.com/dotnet/roslyn/issues/47365
```

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs (hunk 1 above)
Comment (sharwell) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryExpressionParenthesesTests.cs:1]: Correct, the BOM was unintentionally removed in c231e0a61eeba5e14baba9e3c5acef2f17348fae.

**Review by sharwell — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ internal static class CommonParenthesizedExpressionSyntaxExtensions
                 }
             }
 
-            // See https://github.com/dotnet/roslyn/issues/47365
+            // It's not safe to change associativity for dynamic variables as the actual type isn't known. See https://github.com/dotnet/roslyn/issues/47365
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ParenthesizedExpressionSyntaxExtensions.cs:43]: Great thanks!

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by Youssef1313** [comment]
@CyrusNajmabadi @sharwell Anything left to merge this?

**Comment by CyrusNajmabadi** [comment]
@youssef1313 ping me when tests are green. Thanks!

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Tests passes

**Pull request merged**